metaflow/datatools/s3.py

```diff
@@ -1348,32 +1368,228 @@ def _s3op_with_retries(self, mode, **options):
         if self._s3_client_params is not None:
             cmdline.extend(("--s3clientparams", json.dumps(self._s3_client_params)))
 
-        for i in range(S3_RETRY_COUNT + 1):
+        def _get_inject_failures():
+            if mode == "list":
+                return 0
+            return min(90, self._s3_inject_failures)
+
+        retry_count = 0
+        transient_retry_count = 0
+        inject_failures = _get_inject_failures()
+        out_lines = []
+        pending_retries = []
+        loop_count = 0
+        last_ok_count = 0
+        total_ok_count = 0
+
+        def _reset():
+            nonlocal transient_retry_count, inject_failures, out_lines, pending_retries
+            nonlocal loop_count, last_ok_count, total_ok_count
+            transient_retry_count = 0
+            inject_failures = _get_inject_failures()
+            if mode != "put":
+                # For put, even after retries, we keep around whatever we already
+                # uploaded. This is because uploading with overwrite=False is not
+                # an idempotent operation and so some files could be uploaded during
+                # the first try which we should report back.
+                out_lines = []
+            pending_retries = []
+            loop_count = 0
+            last_ok_count = 0
+            total_ok_count = 0  # Reset to zero even if we keep out_lines
+
+        def _update_out_lines(out_lines, ok_lines, resize=False):
+            if resize:
+                # This is the first time around; we make the list big enough. Typically,
+                # there is nothing in out_lines but in some cases (a retry after a
+                # partial result), there may be stuff in it
+                out_lines.extend([None] * (len(ok_lines) - len(out_lines)))
+            for l in ok_lines:
+                idx, rest = l.split(b" ", maxsplit=1)
+                if rest.decode(encoding="utf-8") != TRANSIENT_RETRY_LINE_CONTENT:
+                    out_lines[int(idx.decode(encoding="utf-8"))] = rest
+
+        def _internal(last_ok_count, pending_retries, out_lines, inject_failures):
```

Comment:
Can you make this function name a bit more descriptive and some context to aid in readability?

Reply:
sure, I can do that. The TL;DR was just that I wanted to factor out this from the outer while loop below since it got to be a bit large. I can put it all inline too if that helps.